Fixes breaking bugs in Unity 2019+

stelabouras opened the pull request:
The commits included in this pull request make the asset operational in newer Unity versions (2019+).

stelabouras committed:
The name of the material matches the font name

The material name used to be the same every time ("Font Material").

It now matches the font name so that for every font there's a different
material assigned.

stelabouras committed:
Change TextureImporterType from GUI to Sprite

Changes the `textureType` of `TextureImporter` from `.GUI` to `.Sprite`
so that the texture associated with the font will have a texture type of
Sprite instead of Editor GUI given that's now considered a legacy type.

stelabouras committed:
Disables ReloadFont hack on Unity 2019.2+

This solves an issue when re-importing font assets on Unity 2019.3+
where the font asset became missing on game objects that were using
those fonts.

litefeel commented:
Hello. thanks for your pr.
What happened?
Do you have a demo or video?
Do you have a demo or video?

Thanks.

stelabouras commented:
@litefeel Hey there! Let me preface my message by thanking you for your really useful asset!

So we have been using your asset for quite a while now and we have spotted some issues while using it with the latest Unity versions (2018, 2019+).

More specifically:

* The asset creates a material with the same name of every font ("Font Material") making it hard to distinguish which material controls which font, in case we want to make adjustments to a certain font material. b37cbc17d2ebc93de94d3525a1b1eb3151193994 addresses this issue by setting the name of the material to match the name of the font.
* The call to the `ReloadFont()` method on Unity versions 5.5+ was failing on Unity 2019.3+ as it was resetting the font asset, breaking the links with the elements that were using it, and having the developer to set the font again to all the text elements manually via the Inspector. 1c4cf02618e0734171e33be0f9cebf0a0f2f566a addresses this issue by calling the `ReloadFont()` method only on Unity versions >= 5.5 but < 2019.2 as we have also noticed that it has no effect on Unity 2019.2.
* Finally, the import script was importing the font texture as a `GUI` texture type which looks to be deprecated (Unity Editor lists this type as 'Editor GUI and Legacy GUI', so the `Sprite` type is being used instead on 8dc434b4e53dffb0003acfde3c57cde75ad8bf7e. Maybe some developers will still want to use the `GUI` type, so we can introduce another menu option to "Rebuilt Bitmap Font as GUI" and "Rebuild Bitmap Font as Sprite" if that's preferrable.

litefeel commented:
@stelabouras  Thanks.
Let's talk about it.

- **Material**:  It's a good idea to replace `Font Material` with a font name.
Tip: Text must be set a material only if the font has two or more textures.
- **ReloadFont**: I tested on unity 2018.4 and 2019.3 and it worked fine.
ReloadFont will be break on play mode. So please change to UNITY_2018_4_OR_NEWER.
- **TextureType**: The `Editor GUI and Legacy GUI` have not yet been deprecated. It's work final.
 `Sprite (2D and UI)` is similar to `Editor GUI and Legacy GUI`, but it will generate a sprite to the asset.
If you have to change the type, I prefer `Default`.
Because it does not generate unused assets.

Thank you for your use, I hope you enjoy your use.

litefeel commented on Editor/BFImporter.cs:
```diff
@@ -88,7 +89,9 @@ public static void DoImportBitmapFont(string fntPatn)
-#if UNITY_5_5_OR_NEWER
+#if UNITY_2019_2_OR_NEWER
+            // No-op
+#elif UNITY_5_5_OR_NEWER
```
Please change to UNITY_2018_4_OR_NEWER.
It works fine in unity 2018

stelabouras replied:
Hey, thanks for the suggestion.

So just to be clear, do you suggest changing the `#if` on line 92 or the `#elif` on line 94?

By saying 'it works fine in Unity 2018' do you mean that the `ReloadFont()` call is not needed in Unity 2018.4, so the code becomes:

```suggestion
#if UNITY_2018_4_OR_NEWER
            // No-op
#elif UNITY_5_5_OR_NEWER
```
or that the `ReloadFont()` call is needed for Unity 2018.4, so the code becomes:

```suggestion
#if UNITY_2019_2_OR_NEWER
            // No-op
#elif UNITY_2018_4_OR_NEWER
```

litefeel replied:
```suggestion
#if UNITY_2018_4_OR_NEWER
            // No-op
#elif UNITY_5_5_OR_NEWER
```
It works fine in unity version 2018.4 without `ReloadFont`

stelabouras replied:
Done!

litefeel commented on Editor/BFImporter.cs:
```diff
@@ -112,7 +115,7 @@ private static Texture2D[] DoImportTextures(FntParse parse, string rootPath, Tex
 
                 TextureImporter texImporter = AssetImporter.GetAtPath(texPath) as TextureImporter;
-                texImporter.textureType = TextureImporterType.GUI;
+                texImporter.textureType = TextureImporterType.Sprite;
```
The Editor GUI and Legacy GUI have not yet been deprecated. It's work fine.

stelabouras replied:
This can be reverted, although it won't work in our case where we want the texture type to be respected (we need `Sprite` instead of `GUI` font textures).

Given that there's no way to query the `textureType` out of a `Texture` object (e.g. looking up the already stored `mainTexture` from the material when loading the font in the `DoImportBitmapFont()` method), if we leave this logic as-is, every time we rebuild the bitmap font, it will change the texture type.

What do you think about importing the font texture only if it's not already assigned to the font material?

e.g.
```c#
font.material = material;

if(material.mainTexture == null)
{
    Texture2D[] textures = DoImportTextures(parse, rootPath, fnt);

    material.shader = Shader.Find(textures.Length > 1 ? "BFI/Font" + textures.Length : "UI/Default");

    material.mainTexture = textures[0];

    for (int i = 1; i < textures.Length; i++)
    {
        material.SetTexture("_MainTex" + (i + 1), textures[i]);
    }
}

font.characterInfo = parse.charInfos;
```

litefeel replied:
- get texture type:
~~~
TextureImporter texImporter = AssetImporter.GetAtPath(texPath) as TextureImporter;
var texType = texImporter.textureType;
~~~

- Why use Sprite?
I don’t understand why you have to use Sprite, just because `Editor GUI and Legacy GUI` seems to be deprecated.

stelabouras replied:
> I don’t understand why you have to use Sprite, just because Editor GUI and Legacy GUI seems to be deprecated.

I have reviewed the cases where we used a Sprite type in our project and it looks like this is not needed any more. Maybe we had to switch to that type at some point in the past, but it looks like the `.GUI` type works as well.

A minor suggestion would be to just remove the line where the script sets the explicit `textureType` so that it respects the type set by the user, but it's not a dealbreaker right now; just a suggestion.

stelabouras replied:
I have reverted the change back to the `GUI` type.

litefeel replied:
Thanks.
I will add an option for the texture type on weekend.

litefeel reviewed (changes requested):
Thanks.

stelabouras committed:
Apply merge request suggestions

Applies suggestions on merge request by @litefeel:

* Reverts the `TextureImporterType` back to `GUI`.
* Replaces `UNITY_2019_2_OR_NEWER` with `UNITY_2018_4_OR_NEWER` so that
the `ReloadFont()` method doesn't run on Unity 2018.4+.

litefeel commented:
Hi.
I released 3.1.0
https://github.com/litefeel/Unity-BitmapFontImporter/blob/upm/CHANGELOG.md
